---
title: "Dipteran dsRBP phylogenetics"
output:
  html_document: default
  html_notebook: default
  pdf_document: default
  word_document: default
---
# Setting envinroment
```{r}
args = commandArgs(trailingOnly=TRUE)
system('git config --global user.name "luciorq"')
system('git config --global user.email "luciorqueiroz@gmail.com"')
```
## Testing libraries

## Initial tree
![Flavia's old tree](images/dsRBP_dipterans_v15.PDF)

# Analysis
## Loading libraries
```{r}
source("src/functions.R")
source("lib/rmd2rscript.R")
```
## Initializing variables
List of Files used for the analysis
Here you can change the files used for the analysis, inside 'raw' directory
```{r}
if(length(args)==0) {
  sequences_file = "teste1"
  InterPro = "IPR014720"
  }else{
    sequences_file = args[1]
    InterPro = args[2]
}
tblastn_db = "VB_dipteran_transcripts"
blastx_db = "FB_fly_translated"
```
# VectorBase tBlasn
## Transcripts used for tBlastn
tBlastn was performed against the following libraries:
```{r}
for(line in (readLines(paste0("raw/",tblastn_db,"_list.txt")))){
  print(unlist(strsplit(line, "/"))[length(unlist(strsplit(line, "/")))])
  }
```
### VectorBase sequences for tBlastn
```{r}
DownloadLibraries(tblastn_db)
makeBlastDB(tblastn_db,"nucl")
```
## tBlastn against VectorBase transcripts
```{r}
MLtoSLfasta(paste0("raw/",sequences_file))
Blast("tblastn",paste0("data/",sequences_file,".fasta"), tblastn_db)
```
### Formatting output
```{r}
tblastn_table <- read.csv(paste0("data/",sequences_file,".fasta_alignment.csv"), header = FALSE)
names(tblastn_table) <- c("querySeqId", "subjectSeqId", "queryLength", "alignmentLength", "E-value", "bitscore", "subjectTitle")
tblastn_table$subjectTitle <- unlist(lapply(tblastn_table$subjectTitle, function(x) strsplit(as.character(x), split = "|", fixed = TRUE)[[1]][1]))
tblastn_table <- subset(tblastn_table, (`E-value` < 1e-03))
head(tblastn_table)
```
### Retrieving hit fasta sequences
```{r}
write.table(tblastn_table$subjectSeqId, file = paste0("data/",sequences_file,"_ids.txt"),
            quote = FALSE, row.names = FALSE, col.names = FALSE)
RetrieveFasta(tblastn_db,sequences_file)
```

# FlyBase BlastX
## Translated Proteins used for reverse BlastX against Fruit Flies
```{r}
for(line in (readLines(paste0("raw/",blastx_db,"_list.txt")))){
  print(unlist(strsplit(line, "/"))[length(unlist(strsplit(line, "/")))])
  }
```
### FlyBase sequences for BlastX
```{r}
DownloadLibraries(blastx_db)
makeBlastDB(blastx_db, "prot")
```
## Reverse BlastX against fruit flies
```{r}
#FilterFastaRedundancy(paste0("data/",sequences_file,"_blast_result.fasta"))
Blast("blastx",paste0("data/",sequences_file,"_blast_result.fasta"), blastx_db, outformat = 6)
```
### Formatting output
```{r}
system(paste0("sort -k1,1 -k6,6nr -k5,5n data/",sequences_file,"_blast_result.fasta_alignment.tab | sort -u -k 1,1 --merge > data/",sequences_file,"_filtered.tab"))
blastx_table <- read.table(paste0("data/",sequences_file,"_filtered.tab"), header = FALSE )
#blastx_table <- read.table(paste0("data/",sequences_file,"_blast_result.fasta_alignment.tab"), header = FALSE )
blastx_table$V7 <- blastx_table$V10
names(blastx_table) <- c("querySeqId", "subjectSeqId", "queryLength", "alignmentLength", "E-value", "bitscore", "subjectTitle")
blastx_table <- blastx_table[,1:7]
head(blastx_table)
```
### Formatting table
```{r}
#FormatTable()
reverse_blast_list = c()
reverse_evalue_list = c()
for(transcript in tblastn_table$subjectSeqId){
  i = 1
  for(hit in blastx_table$querySeqId){
    if(hit == transcript){
      hit_name = as.character(blastx_table$subjectTitle[i])
      reverse_blast_list = c(reverse_blast_list, hit_name)
      reverse_evalue = blastx_table$`E-value`[i]
      reverse_evalue_list = c(reverse_evalue_list, reverse_evalue)
    }
    i = i + 1
  }
}

final_document <- tblastn_table[,c("querySeqId","subjectSeqId","E-value","subjectTitle")]
#final_document$reverseBlast <- reverse_blast_list
#final_document$reverseBlastEvalue <- reverse_evalue_list
final_document <- subset(final_document, (`E-value` < 1e-03))
write.csv(final_document, file = paste0("results/",sequences_file,".csv"), row.names = FALSE)
final_document
```

### Retrieving transcripts fasta sequences
```{r}
write.table(final_document$subjectSeqId, file = paste0("data/",sequences_file,"_nucl_ids.txt"),
            quote = FALSE, row.names = FALSE, col.names = FALSE)
RetrieveFasta(tblastn_db,paste0(sequences_file,"_nucl"))
paste0("data/",sequences_file,"_nucl_blast_result.fasta")
```
## Finding dsRBP domain in ORFs from transcripts
```{r}
GetORF(sequences_file)
```
### Extracting domain information from ORFs
```{r}
RunIprScan((paste0("data/",sequences_file,"_orf.fasta")), 
           (paste0("data/iprscan_results/",sequences_file,"_orf")))
```
### Retrieving ORFs with `r InterPro` domain
```{r}
ORFwithDomain(sequences_file, InterPro)
```

### Concatenating novel translated transcripts to the 'raw' data
```{r}
system(paste0("cat data/",sequences_file,".fasta > data/",sequences_file,"_result.fasta"))
system(paste0("cat data/",sequences_file,"_domain_filtered_ORF.fasta >> data/",sequences_file,"_result.fasta"))
```
# INTERPRO Scan
## Batch InterProScan for `r sequences_file`
```{r}
#RunIprScan((paste0("data/",sequences_file,"_result.fasta")), 
#           (paste0("data/iprscan_results/",sequences_file)))
```

# Building Trees
## MAFFT alignment
```{r}
FilterFastaRedundancy(paste0("data/",sequences_file,"_result.fasta"),paste0("data/",sequences_file,"_result_nr.fasta") )
Alignment("mafft","--auto", sequences_file)
```
##
```{r}
BuildTrees(sequences_file, "Ce.Rde-4")
```

# Converting this notebook to an executable script
```{r}
rmd2rscript("dsRBP.Rmd")
```


